---
title: "Introduction to radmixture"
author: "Beilei Bian"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

This is an introduction to radmixture which could help you with estimating individual ancestries from large SNP genotype data. 

This document introduces you how to use this package. 

### Installation

- Install the latest development version from github with
```r
devtools::install_github("wegene-llc/radmixture")
```
- Install the latest version from CRAN
```r
install.packages("radmixture")
```

First, you must prepare your raw data file as follow:

![](https://cloud.githubusercontent.com/assets/18478302/22725657/b4abd21a-ee09-11e6-9ef8-a4092be538e8.png)

and read it into R with `read.table` or `read.csv`.

```r
library(radmixture)
genotype <- read.table(file = '/path/to/file')
# genotype <- read.csv(file = 'path/to/file')
```

## Calculate your ancestry components with public dataset

### Example

- Use `tfrdpub()`to transfer your raw data to a format understood by radmixture

```r
res <- tfrdpub(genotype, 13, global13.alleles, global13.13.F)
```

- Calculate

```r
ances <- fFixQN(res$g, res$q, res$f, tol = 1e-4, method = 'BR', pubdata = "K13")
```
`ances$q` is your K13 result.

### References

- D.H. Alexander, J. Novembre, and K. Lange. Fast model-based estimation of ancestry in unrelated individuals. Genome Research, 19:1655â€“1664, 2009.
- H. Zhou, D. H. Alexander, and K.  Lange. A quasi-Newton method for accelerating the convergence of iterative optimization algorithms. Statistics and Computing, 2009.

### Contributions
We welcome contributions on radmixture. You can fork this repo and make your changes and submit a pull request.
Report bug issues on [issue page](https://github.com/wegene-llc/radmixture/issues).

### License

MIT + LICENSE

### TODO

- Add tutorial for radmixture